devito/operator/operator.py

```diff
@@ -1181,10 +1211,15 @@ def rcompile(expressions, kwargs, options, target=None):
     # Recursive profiling not supported -- would be a complete mess
     kwargs.pop('profiler', None)
 
-    return cls._lower(expressions, **kwargs)
+    # Recursive compilation is expensive, so we cache the result because sometimes
+    # it is called multiple times for the same input
+    compiled = RCompiles(expressions, cls)
+    compiled(**kwargs)
+
+    return compiled.output
```

Comment:
you could avoid the `__call__` which seems dangerous considering that's what CacheInstances uses to catch and instead just tp `compiled.output(**kwargs)`

Reply:
done